qiskit/algorithms/linear_solvers/matrices/tridiagonal.py

```diff
@@ -0,0 +1,210 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Hamiltonian simulation of tridiagonal Toeplitz symmetric matrices."""
+
+from typing import Optional
+
+import numpy as np
+from scipy.sparse import diags
+
+from qiskit.circuit import QuantumCircuit, QuantumRegister, AncillaRegister, Qubit
+
+
+class Tridiagonal(QuantumCircuit):
+    """Class of tridiagonal Toeplitz symmetric matrices"""
+
+    def __init__(self, num_state_qubits: int, main_diag: float, off_diag: float,
+                 tolerance: Optional[float] = None, time: Optional[float] = None,
+                 trotter: Optional[int] = 1, name: str = 'tridi') -> None:
+
+        qr_state = QuantumRegister(num_state_qubits)
+        if num_state_qubits > 1:
+            qr_ancilla = AncillaRegister(max(1, num_state_qubits - 1))
+            super().__init__(qr_state, qr_ancilla, name=name)
+        else:
+            super().__init__(qr_state, name=name)
+
+        self._num_state_qubits = None
+
+        self._main_diag = main_diag
+        self._off_diag = off_diag
+        self._tolerance = tolerance if tolerance is not None else 1e-2
+        self._time = time if time is not None else 1
+        self._trotter = trotter
+
+        self._num_state_qubits = num_state_qubits
+
+    def set_simulation_params(self, time: float, tolerance: float):
+        """Override the simulation parameters"""
+        self.tolerance(tolerance)
+        self.time(time)
+
+    def tolerance(self, tolerance: float):
+        """Return the error tolerance"""
+        self._tolerance = tolerance
+
+    def time(self, time: float):
+        """Return the time of the evolution"""
+        self._time = time
+        # Update the number of trotter steps. Max 7 for now, upper bounds too loose.
+        self._trotter = min(self._num_state_qubits + 1,
```

Comment:
I imagine the comments for this were updated in the getter/setter change since it talks about returning time but is setting things. Why does this also have the 'side-effect' of setting trotter steps. From the constructor they seem to be independent parameters. If this is intentional, and still done, then it should clearly be stated in docstring.

Reply:
Yes, the comments were changed with the getter/setter update. And updated the docstring - the error tolerance we allow in the approximation for the hamiltonian simulation of tridiagonal matrices is a function of the evolution time (`time` in this version of the code) and the number of trotter steps.

Reply:
I noted the 'side-effect' of trotter steps being updated too here since in the constructor they (evo_time and trotter) are apparently independent parameters I can pass. Yet here, if I set time this way, it overrides whatever trotter steps were passed in. There seems therefore to be a disconnect to me between construction and this method .

Maybe if num trotter steps is going to be auto computed from tolerance and evo time it should not be passable via constructor. Or maybe one could allow it to be passed independently if say it and tolerance were allowed to be None where you would either have to set tolerance or num trotter steps (clearly if setting both one of them would have to take precedence).